src/SymbolCollector.Core/Client.cs

```diff
@@ -107,62 +114,138 @@ private async Task UploadAsync(string debugId, string file, CancellationToken ca
+                getBuildId = GetMachOBuildId;
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                getBuildId = GetElfBuildId;
+            }
+            else
+            {
+                // TODO: This needs to be check at the start of the program though
+                throw new NotSupportedException($"OS {RuntimeInformation.OSDescription} is not supported.");
+            }
+
             foreach (var file in files)
             {
                 _logger.LogInformation("Processing file: {file}.", file);
-                IELF? elf = null;
-                string? buildIdHex;
-                try
+
+                var buildId = getBuildId(file);
+                if (buildId is null)
                 {
-                    // TODO: find an async API
-                    if (!ELFReader.TryLoad(file, out elf))
+                    if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
                     {
-                        _logger.LogWarning("Couldn't load': {file} with ELF reader.", file);
-                        continue;
-                    }
-
-                    var hasBuildId = elf.TryGetSection(".note.gnu.build-id", out var buildId);
-                    if (!hasBuildId)
-                    {
-                        _logger.LogWarning("No Debug Id in {file}", file);
-                        continue;
+                        // Check if it's a Fat Mach-O
+                        if (_fatBinaryReader.TryLoad(file, out var test) && test is { } fatMachO)
+                        {
+                            _logger.LogInformation("Fat binary file with {count} Mach-O files: {file}.", fatMachO.Header.FatArchCount, file);
+
+                            using (_logger.BeginScope(new {FatMachO = file}))
+                            using (fatMachO)
+                            {
+                                foreach (var buildIdFile in GetFiles(fatMachO.MachOFiles))
+                                {
+                                    yield return buildIdFile;
+                                }
+                            }
+                        }
                     }
+                    continue;
+                }
 
-                    var hasUnwindingInfo = elf.TryGetSection(".eh_frame", out _);
-                    var hasDwarfDebugInfo = elf.TryGetSection(".debug_frame", out _);
+                yield return (buildId, file);
+            }
+        }
 
-                    if (!hasUnwindingInfo && !hasDwarfDebugInfo)
+        private string? GetMachOBuildId(string file)
+        {
+            try
+            {
+                // TODO: find an async API if this is used by the server
+                if (MachOReader.TryLoad(file, out _) == MachOResult.OK)
+                {
+                    using var algorithm = SHA256.Create();
+                    var hash = algorithm.ComputeHash(File.ReadAllBytes(file));
```

Comment:
This will still need to look for the LC_VERSION load command.

Reply:
Assuming you mean: `LC_VERSION_MIN_MACOSX` and `LC_VERSION_MIN_IPHONEOS`